core/sql/sort/Topn.cpp

```diff
@@ -0,0 +1,337 @@
+             CollHeap* heap, SortError* sorterror, Lng32 explainNodeId, SortUtil* sortutil):
+             SortAlgo(runsize, recsize, doNotallocRec, keysize, scratch, explainNodeId),
+             loopIndex_(0), heap_(heap), sortError_(sorterror),
+             sortUtil_(sortutil)
+{
+   //runsize is TopN size. Fixed.
+   allocRunSize_ = runsize;
+   
+   //Actual run size after all elements read.
+   runSize_ = 0;   
+   
+   isHeapified_ = FALSE;
+   
+  //allocateMemory failureIsFatal is defaulted to TRUE means allocation failure results in 
+  //longjump to jump handler defined in ex_sort.cpp. Only applicable on NSK.
+  topNKeys_  = (RecKeyBuffer *)heap_->allocateMemory(sizeof(RecKeyBuffer) * allocRunSize_);  
+ 
+  // Below asserts useful in debug mode. Also asserts if longjmp did not happen.
+  //ex_assert(rootRecord_!= NULL, "Sort: Initial rootRecord_ allocation failed"); 
+  ex_assert(topNKeys_  != NULL, "Sort: Initial topNKeys_ allocation failed");  
+
+  recNum_ = 0;
+  ExOperStats *stat = sortUtil_->config()->getCallingTcb()->getStatsEntry();
+  if (stat)
+    bmoStats_ = stat->castToExBMOStats();
+  else
+    bmoStats_ = NULL;
+  if (bmoStats_)
+    bmoStats_->updateBMOHeapUsage((NAHeap *)heap_);
+}
+
+ 
+//------------------------------------------------------------------------
+// Class Destructor: Delete all the heap space pointed by pointers in Qsort
+//------------------------------------------------------------------------
+TopN::~TopN(void)
+{
+
+  for (int i = 0; i < runSize_; i++)
+	  topNKeys_[i].rec_->releaseTupp();
+
+  
+  if (topNKeys_ != NULL) {
+    NADELETEBASIC(topNKeys_, heap_);
+    topNKeys_ = NULL;
+  }
+  if (bmoStats_)
+    bmoStats_->updateBMOHeapUsage((NAHeap *)heap_);
+
+}
+
+Lng32 TopN::sortSend(void *rec, ULng32 len, void* tupp)		                        
+{
+  //if heap not built means, TopN array has more slots 
+  //available to fill. 
+  if(! isHeapified_)
+  {
+	  ex_assert(loopIndex_ >= 0, "TopN::sortSend: loopIndex_ is < 0");
+	  ex_assert(loopIndex_ < allocRunSize_, "TopN::sortSend: loopIndex_ > allocRunSize_");
+	  ex_assert(rec != NULL, "TopN::sortSend: rec is NULL");
+	
+	  Record * insertRec = new Record(rec, len, tupp, heap_, sortError_);
+	  
+	  topNKeys_[loopIndex_].key_ = insertRec->extractKey(keySize_, sortUtil_->config()->numberOfBytesForRecordSize());
+	  topNKeys_[loopIndex_].rec_ = insertRec;
+	  if (++loopIndex_  == allocRunSize_)
+	  {
+		 //Reaching here means, we have filled up the array. 
+		 //Now heapify the array to start accepting/eliminating new elements from now on.
+		 
+         //Note lookIndex_ contains the current number of filled elements.
+		  
+		 buildHeap();
+	  }
+	  return SORT_SUCCESS;
+  }
+  
+  //Reaching here means, heap is already build. 
+  //Check if the new rec belongs to this heap by comparing the
+  //new rec key with the root node of the heap ( root node is always the greatest).
+  insertRec(rec, len, tupp);
```

Comment:
Odd that this method name has the same name as the variable declared at line 121. Doesn't hurt anything but does make the code a bit more complicated to understand. And it might break if you reorder the code later.